### PR TITLE
MEN-3187: Overview/Update-Modules

### PR DESCRIPTION
**Pull request by oleorhagen**
Mostly transferred clean for our old docs. Some minor wordings here and there.

### CODE REVIEW

File: 02.Overview/07.Mender-Update-Module/docs.md [hunk 1]
```diff
@@ -4,3 +4,302 @@ taxonomy:
...
+
+The built-in update mechanism in Mender is [dual rootfs
+updates](../../architecture/overview#robust-updates). To do other types of
+updates, you will need an _Update module_ for your device.
+
+An _Update Module_ extends the Mender client to support a new type of software
+update, such as a package manager, container, bootloader or even updates of
+attached microcontrollers. Tailor the Update Module to a specific use case or
+environment, such as a proprietary bootloader, or install a set of `.deb`
+packages.
```
Comment (kacf) [inline 02.Overview/07.Mender-Update-Module/docs.md:18]: Suggestion: "**Update Modules can be tailored** to a specific use case or environment, such as **updating** a proprietary bootloader, or **installing** a set of `.deb` packages."

File: 02.Overview/07.Mender-Update-Module/docs.md [hunk 2]
```diff
@@ -4,3 +4,302 @@ taxonomy:
...
+and container-based updates are available and documented together with
+community-supported Update Modules at [Mender
+Hub](https://hub.mender.io?target=_blank).
+
+This document introduces how Update Modules work, so you can develop your own
+Update Module that meets your specific needs.
+
+An Update Module implements one or more of the actions that the Mender client
+carries out when installing an Artifact payload. The *core* action all Update
+Modules must implement is the Install action, which implements how an Artifact
```
Comment (kacf) [inline 02.Overview/07.Mender-Update-Module/docs.md:30]: Perhaps we should start right off the bat with calling it `ArtifactInstall`, so there's no confusion.

File: 02.Overview/07.Mender-Update-Module/docs.md [hunk 3]
```diff
@@ -4,3 +4,302 @@ taxonomy:
...
+
+Update Modules are modeled after the same execution flow as [state
+scripts](../../artifacts/state-scripts). For the development of Update Modules
+it is important to have a basic understanding of the workflow.
+
+![Update Modules state machine](update-modules-state-machine.png)
+
+The most relevant states for developing an Update Module are:
+
+* `ArtifactInstall` state should installs the update into its final destination.
```
Comment (kacf) [inline 02.Overview/07.Mender-Update-Module/docs.md:45]: `s/installs/install/`

File: 02.Overview/07.Mender-Update-Module/docs.md [hunk 4]
```diff
@@ -4,3 +4,302 @@ taxonomy:
...
+* `ArtifactCommit` state should makes the update permanent, in cases where it is still possible to roll back.
+* `ArtifactRollback` state should rolls back to the previously known good state, normally by restoring some kind of backup.
```
Comment (kacf) [inline 02.Overview/07.Mender-Update-Module/docs.md:47]: `s/rolls/roll/`

File: 02.Overview/07.Mender-Update-Module/docs.md [hunk 5]
```diff
@@ -4,3 +4,302 @@ taxonomy:
...
+
+### Create the Update Module script
+
+In the device terminal, go to `/usr/share/mender/modules/v3`
+
+```bash
+cd /usr/share/mender/modules/v3
+```
+
+and create an script named `web-file` with the following content:
```
Comment (kacf) [inline 02.Overview/07.Mender-Update-Module/docs.md:99]: Incorrect instructions. It's telling the user to create a file with the content of the below snippet, but the snippet contains commands to create the file, not just the content. Should be changed to either one or the other: Provide just the content, or instruct the user to execute the snippet.

File: 02.Overview/07.Mender-Update-Module/docs.md [hunk 6]
```diff
@@ -4,3 +4,302 @@ taxonomy:
...
+Then create a new Mender Artifact with this as the only payload, for our new
+`web-file` Update Module:
+
+```bash
+./mender-artifact write module-image -t $DEVICE_TYPE  -o ~/Downloads/web-file-1.mender -T web-file -n web-file-1.0 -f hello-world
+```
+The command line options are detailed below:
+* `-t` - The compatible device type of this Mender Artifact.
+* `-o` - The path where to place the output Mender Artifact. This should always have a .mender suffix.
+* `-T` - The update module name. The name should be same of the update module script which is present inside `/usr/share/mender/modules/v3` path.
```
Comment (kacf) [inline 02.Overview/07.Mender-Update-Module/docs.md:161]: * `s/same of/same as/`
* Change all "update module" instances to "Update Module"

File: 02.Overview/07.Mender-Update-Module/docs.md [hunk 7]
```diff
@@ -4,3 +4,302 @@ taxonomy:
...
+The `mender-artifact` tool allows multiple specification of files using the `-f`
+flags. This means we can create an Artifact with multiple files like this:
+
+```bash
+./mender-artifact write module-image -t $DEVICE_TYPE  -o ~/Downloads/web-file-1.mender -T web-file -n web-file-1.1 -f my-file-1 -f my-file-2 -f my-file-3 ...
+```
+The command line options are detailed below:
+* `-t` - The compatible device type of this Mender Artifact.
+* `-o` - The path where to place the output Mender Artifact. This should always have a .mender suffix.
+* `-T` - The update module name. The name should be same of the update module script which is present inside `/usr/share/mender/modules/v3` path.
```
Comment (kacf) [inline 02.Overview/07.Mender-Update-Module/docs.md:212]: Same here.

File: 02.Overview/07.Mender-Update-Module/docs.md [hunk 8]
```diff
@@ -4,3 +4,302 @@ taxonomy:
...
+For an Update Module to support Rollback, two additions are required in the script:
+* Return "Yes" on `SupportsRollback` action call
+* Implement a mechanism that will save a "known good state" before the installation in `ArtifactInstall`, and restore that state in `ArtifactRollback`.
+
+See below the updated version of our `web-file` Update Module:
+
+! The updated version of the Update Module will remove all files in `/var/www`. It is meant to showcase the usage of the feature but can be unsafe in real environments.
+
+```bash
+cat << "EOF" > web-file
```
Comment (kacf) [inline 02.Overview/07.Mender-Update-Module/docs.md:236]: Consider changing this snippet too, either specify that it is pure content, or to be executed.

File: 02.Overview/07.Mender-Update-Module/docs.md (hunk 1 above)
Comment (oleorhagen) [reply 02.Overview/07.Mender-Update-Module/docs.md:18]: Passive voice again though, but are you thinking about the highlighting?

File: 02.Overview/07.Mender-Update-Module/docs.md (hunk 5 above)
Comment (oleorhagen) [reply 02.Overview/07.Mender-Update-Module/docs.md:99]: I've been staring at this for 2 minutes now. Where exactly in the script does the script get created?

File: 02.Overview/07.Mender-Update-Module/docs.md [hunk 9]
```diff
@@ -4,3 +4,302 @@ taxonomy:
     category: docs
     label: reference
 ---
+
+## Introduction
+
+The built-in update mechanism in Mender is [dual rootfs
+updates](../../architecture/overview#robust-updates). To do other types of
```
Comment (drewmoseley) [inline 02.Overview/07.Mender-Update-Module/docs.md:11]: mechanism in Mender is a [dual rootfs update]....

File: 02.Overview/07.Mender-Update-Module/docs.md (hunk 1 above)
Comment (drewmoseley) [reply 02.Overview/07.Mender-Update-Module/docs.md:18]: Yeah, I think "Developers can tailor Update Modules ...."

File: 02.Overview/07.Mender-Update-Module/docs.md [hunk 10]
```diff
@@ -4,3 +4,302 @@ taxonomy:
...
+An _Update Module_ extends the Mender client to support a new type of software
+update, such as a package manager, container, bootloader or even updates of
+attached microcontrollers. Tailor the Update Module to a specific use case or
+environment, such as a proprietary bootloader, or install a set of `.deb`
+packages.
+
+General-purpose Update Modules, including support for file-based, package based
+and container-based updates are available and documented together with
+community-supported Update Modules at [Mender
+Hub](https://hub.mender.io?target=_blank).
```
Comment (drewmoseley) [inline 02.Overview/07.Mender-Update-Module/docs.md:23]: Should this link point to https://hub.mender.io/c/update-modules/13 instead?

File: 02.Overview/07.Mender-Update-Module/docs.md [hunk 11]
```diff
@@ -4,3 +4,302 @@ taxonomy:
...
+
+This document introduces how Update Modules work, so you can develop your own
+Update Module that meets your specific needs.
+
+An Update Module implements one or more of the actions that the Mender client
+carries out when installing an Artifact payload. The *core* action all Update
+Modules must implement is the Install action, which implements how an Artifact
+payload is installed on the device. However, there are other actions that may be
+implemented depending on the desired functionality and use case for an Update
+Module, such as Rollback.
```
Comment (drewmoseley) [inline 02.Overview/07.Mender-Update-Module/docs.md:33]: no need for the comma here.

File: 02.Overview/07.Mender-Update-Module/docs.md (hunk 3 above)
Comment (drewmoseley) [inline 02.Overview/07.Mender-Update-Module/docs.md:45]: I think all these bullets should read:
blah state installs (or other verb) the ...

File: 02.Overview/07.Mender-Update-Module/docs.md [hunk 12]
```diff
@@ -4,3 +4,302 @@ taxonomy:
...
+![Update Modules state machine](update-modules-state-machine.png)
+
+The most relevant states for developing an Update Module are:
+
+* `ArtifactInstall` state should installs the update into its final destination.
+* `ArtifactCommit` state should makes the update permanent, in cases where it is still possible to roll back.
+* `ArtifactRollback` state should rolls back to the previously known good state, normally by restoring some kind of backup.
+
+Every state is optional for an Update Module to implement, however in practice
+all Update Modules would likely start with implementing the `ArtifactInstall`
```
Comment (drewmoseley) [inline 02.Overview/07.Mender-Update-Module/docs.md:50]: all Update Modules will implement the Artifact Install state.

File: 02.Overview/07.Mender-Update-Module/docs.md [hunk 13]
```diff
@@ -4,3 +4,302 @@ taxonomy:
...
+
+You will need a device with a Mender client installed. For development purposes
+you also need shell access to the device (e.g. via SSH).
+
+To install the Mender client on your device, follow the instructions in the
+[Installing](../../client-configuration/installing) documentation.
+
+### Workstation with mender-artifact
+
+We will use the `mender-artifact` tool to create the payloads required for
```
Comment (drewmoseley) [inline 02.Overview/07.Mender-Update-Module/docs.md:80]: create the artifacts required...

File: 02.Overview/07.Mender-Update-Module/docs.md (hunk 1 above)
Comment (kacf) [reply 02.Overview/07.Mender-Update-Module/docs.md:18]: Yes, Drew's suggestion also works. My point was that I think there should be a "can" in there, and the two last highlighted words need the correct tense.

File: 02.Overview/07.Mender-Update-Module/docs.md (hunk 5 above)
Comment (kacf) [reply 02.Overview/07.Mender-Update-Module/docs.md:99]: See the top:

```bash
cat << "EOF" > web-file
```

Comment (oleorhagen) [reply 02.Overview/07.Mender-Update-Module/docs.md:99]: Ahh, I understand now :1st_place_medal:

**Comment by oleorhagen** [comment]
@drewmoseley should be good now

**Comment by drewmoseley** [comment]
LGTM. (I don't see an approve changes link but consider this my approval.)

**Pull request merged**